### DISCUSSION TITLE
Export PDF/A

### DISCUSSION THREAD

**Discussion by nathan30** [py-pdf/pypdf#700] [Q&A]
Hi,

Is it possible to export PDF using the PDF/A standard ?

Thanks in advance

**Comment by geimist** [comment]
I also miss this feature very much.

**Comment by MartinThoma** [comment]
I think adding this feature is a pretty hard task.

**Comment by geimist** [comment]
Thank you for your quick reply.
Do you know an alternative way (Linux, commandline)? I only find recommendations with ghostscript, but the users of my software don't have ghostscript on their NAS devices.
I want to write metadata to a PDF/A with PyPDF2 and after that it is only a PDF in version 1.3.

**Reply by MartinThoma** [reply]
> I want to write metadata to a PDF/A with PyPDF2 and after that it is only a PDF in version 1.3.

PDF 1.3 can be a valid PDF/A. You might want to read [What does “PDF/A compliant” mean?](https://developers.foxit.com/pdf/guide-pdfa-compliance/)

**Reply by MartinThoma** [reply]
Thinking about it, we could potentially do this iteratively. There are a few easy ones (e.g. removing passwords, adding a metadata flag). We could add it with a warning that it's not complete and work on it.

What do you think would be a reasonable interface? `writer.make_pdfa_compliant()`?

I vaguely remember that there were a few things that could not automatically be checked, but I might be wrong about that.

**Reply by geimist** [reply]
> PDF 1.3 can be a valid PDF/A. You might want to read [What does “PDF/A compliant” mean?](https://developers.foxit.com/pdf/guide-pdfa-compliance/)

I assume that all source documents are already valid PDF/A documents which have been output by OCRmyPDF in the previous step. These documents are also correctly recognized as PDF/A (1.7) by Acrobat, for example, and warn about a change. If I set the metadata with PyPDF2, the PDF documents lose this attribute. However, various users of my [tool](https://git.geimist.eu/geimist/synOCR/src/branch/master/README.md) rely on the correct attribute.

My suggestion would be: 
If the source files are PDF/A, they can be marked as such again if desired (with a note that the source files have not been checked).

> What do you think would be a reasonable interface? `writer.make_pdfa_compliant()`?

Sorry, but unfortunately I have no knowledge of Python.
For the metadata, I use PyPDF2 (in Bash) as follows:
```bash
py_meta="'/Author': '$document_author',"
py_meta="$(printf "$py_meta\n'/Keywords': \'${CommaSeparatedKeywords}\',")"
py_meta="$(printf "$py_meta\n'/CreationDate': \'D:${creationDate}\'")"


{   echo "import pprint"

    echo "from PyPDF2 import PdfFileReader, PdfFileMerger"

    echo "if __name__ == '__main__':"
    echo "    file_in = open('$outputtmp', 'rb')"
    echo "    pdf_reader = PdfFileReader(file_in)"
    echo "    metadata = pdf_reader.getDocumentInfo()"

    echo "    pdf_merger = PdfFileMerger()"
    echo "    pdf_merger.append(file_in)"
    echo "    pdf_merger.addMetadata({$py_meta})"
    echo "    file_out = open('$outputtmpMeta', 'wb')"
    echo "    pdf_merger.write(file_out)"

    echo "    file_in.close()"
    echo "    file_out.close()"
} | python3
```

**Reply by MartinThoma** [reply]
Ah, now I understand your problem better - and that should be a way simpler issue!

I'm sorry that I was / am a bit grumpy at the moment. I have attic apartment which is currently over 35°C and even at night doesn't go under ~28°C for a week now... that's just raising my stress level quite a bit.

I've opened https://github.com/py-pdf/PyPDF2/issues/1012 so that we can start working on the first bit. I think the first bit should be that if we "merge" only a single PDF/A compliant document, it should stay PDF/A compliant.

**Reply by MartinThoma** [reply]
As an unrelated side-note, I would recommend to adjust your script a tiny bit:

```
py_meta="'/Author': '$document_author',"
py_meta="$(printf "$py_meta\n'/Keywords': \'${CommaSeparatedKeywords}\',")"
py_meta="$(printf "$py_meta\n'/CreationDate': \'D:${creationDate}\'")"


{   echo "import pprint"

    echo "from PyPDF2 import PdfReader, PdfMerger"

    echo "if __name__ == '__main__':"
    echo "    reader = PdfReader('$outputtmp')"
    echo "    metadata = reader.metadata"

    echo "    merger = PdfMerger()"
    echo "    merger.append(reader)"
    echo "    merger.add_metadata({$py_meta})"
    echo "    with open('$outputtmpMeta', 'wb') as fp:"
    echo "        merger.write(fp)"
} | python3
```

We had a couple of naming changes recently in PyPDF2. The old names (PdfFileReader, PdfFileMerger) will still work for a while, but moving is better (PdfReader, PdfMerger).

**Reply by geimist** [reply]
> I've opened #1012 so that we can start working on the first bit. I think the first bit should be that if we "merge" only a single PDF/A compliant document, it should stay PDF/A compliant.

Thank you for your work and for taking care of the problem.

> As an unrelated side-note, I would recommend to adjust your script a tiny bit:

Thanks for this tip too!

**Reply by manathan1984** [reply]
Thanks for working on this! Is there a way to simply mark "manually" through PyPDF2 a pdf as pdf/a?
